src/github.com/getlantern/peerscanner/host.go

```diff
@@ -0,0 +1,357 @@
+package main
+
+import (
+	"crypto/tls"
+	"fmt"
+	"io"
+	"math/rand"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/getlantern/cloudflare"
+	"github.com/getlantern/enproxy"
+	"github.com/getlantern/tlsdialer"
+)
+
+var (
+	// Set a short ttl on DNS entries
+	ttl = 30 * time.Second
+
+	// Test with a period of half the ttl
+	testPeriod = ttl / 2
+
+	// If we haven't had a successul test or reset after this amount of time,
+	// terminate.
+	terminateAfter = 10 * time.Minute
+
+	dialTimeout    = 3 * time.Second // how long to wait on connecting to host
+	requestTimeout = 6 * time.Second // how long to wait for test requests to process
+	proxyAttempts  = 3               // how many times to try a test request before considering host down
+
+	// Sites to use for testing connectivity. WARNING - these should only be
+	// sites with consistent fast response times, around the world, otherwise
+	// tests may time out.
+	testSites = []string{"www.google.com", "www.youtube.com", "www.facebook.com"}
+)
+
+type status struct {
+	online            bool
+	connectionRefused bool
+}
+
+// host is an actor that represents a host entry in CloudFlare and is
+// responsible for checking connectivity to the host and updating CloudFlare DNS
+// accordingly.
+type host struct {
+	name   string
+	ip     string
+	record *cloudflare.Record
+	groups map[string]*group
+
+	resetCh      chan string
+	unregisterCh chan interface{}
+	statusCh     chan chan *status
+
+	proxiedClient *http.Client
+}
+
+func (h *host) String() string {
+	return fmt.Sprintf("%v (%v)", h.name, h.ip)
+}
+
+// newHost creates a new host for the given name, ip and optional DNS record.
+func newHost(name string, ip string, record *cloudflare.Record) *host {
```

Comment:
This is the run loop for each host actor.  This is the most critical logic to understand.